### ISSUE TITLE
Generics

### ISSUE THREAD

**Issue by andymc12** [eclipse/microprofile-graphql#84]
Per conversation in gitter and previously in the MP GraphQL meetings, we discussed the issues surrounding the use of generics as a return type or parameter in a query/mutation.

The consensus is that collections (i.e. `List<Book>`, `Set<Widget>`, etc.) should work just fine.  Naming of types/input types for other generics are awkward, and thus allowed, but not recommended - for example:
```
@Type("Box")
@InputType("BoxInput")
class Container<T> {...}
```

```
@Query
public Container<Book> findBooks() {...}

@Mutation
public Container<Book> saveBooks(Container<Book> input) {...}
```

The schema might generate something like:
```
type Box_Book { ... }
input Box_BookInput { ... }
```

with no way for a user to override it unless they created a concrete class like:
```
@Type("BookBox")
@InputType("BookBoxInput")
class MyBookBox extends Container<Book> { ... }
```

This should be the recommended approach for dealing with non-collection generic types.

Further, we should also document that inferred generics are not allowed to be the return type for query/mutation methods - i.e.
```
@Query public <T> T something() {...}
```

**Comment by kaqqao** [comment]
What happens if we have `Container<Box>` as an input, where `Container` has `@InputType("BoxInput")` and `Book` has `@InputType("BookInput")`?
Is the total name `BoxInput_BookInput`?

**Comment by andymc12** [comment]
Resolved by pull request #129
